crates/core/src/metadata_store/providers/objstore/immutable_store.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+
+use bytes::Bytes;
+use bytestring::ByteString;
+
+use crate::metadata_store::{Precondition, VersionedValue, WriteError};
+use restate_types::errors::GenericError;
+use restate_types::Version;
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
+pub(crate) struct ImmutableStore {
+    /// a version that is increased with every mutating operation.
+    /// do not confuse with the per-key versions.
+    global_version: Version,
+    /// The collection of versioned values.
```

Comment:
Would we not be better off doing this per metadata store key, rather than as a monolithic blob? We don't have a global linearizability requirement and writing key-value pairs as individual objects would reduce conflicts. Not too worried about the value sizes but there's still some nontrivial write amplification going on with the current implementation.